src/main/rx/nrf24_cx10.c

```diff
@@ -0,0 +1,369 @@
+
+    NRF24L01_SetRxMode(); // enter receive mode to start listening for packets
+#else
+    XN297_SetTXAddr(rxTxAddr, RX_TX_ADDR_LEN); // must do before SetRXAddr, even for RX only
+    XN297_SetRXAddr(rxTxAddr, RX_TX_ADDR_LEN);
+    NRF24L01_FlushRx();
+    NRF24L01_WriteReg(NRF24L01_11_RX_PW_P0, payloadSize + 2); // payload + 2 bytes CRC
+    NRF24L01_SetRxMode();
+    NRF24L01_WriteReg(NRF24L01_1C_DYNPD, 0x00); // Disable dynamic payload length on all pipes
+#endif
+}
+
+/*
+ * Returns true if it is a bind packet.
+ */
+STATIC_UNIT_TESTED bool checkBindPacket(const uint8_t *packet)
+{
+    const bool bindPacket = (packet[0] == 0xaa);
+    if (bindPacket) {
+        txId[0] = packet[1];
+        txId[1] = packet[2];
+        txId[2] = packet[3];
+        txId[3] = packet[4];
+        return true;
+    }
+    return false;
+}
+
+STATIC_UNIT_TESTED uint16_t convertToPwmUnsigned(const uint8_t* pVal)
+{
+    uint16_t ret = (*(pVal + 1)) & 0x7f; // mask out top bit which is used for a flag for the rudder
+    ret = (ret << 8) | *pVal;
+    return ret;
+}
+
+void cx10SetRcDataFromPayload(uint16_t *rcData, const uint8_t *payload)
+{
+    const uint8_t offset = (cx10Protocol == NRF24RX_CX10) ? 0 : 4;
+    rcData[NRF24_ROLL] = convertToPwmUnsigned(&payload[5 + offset]);  // aileron
+    rcData[NRF24_PITCH] = convertToPwmUnsigned(&payload[7 + offset]); // elevator
+    rcData[NRF24_THROTTLE] = convertToPwmUnsigned(&payload[9 + offset]);
+    rcData[NRF24_YAW] = convertToPwmUnsigned(&payload[11 + offset]);  // rudder
+}
+
+static void hopToNextChannel(void)
+{
+    ++rfChannelIndex;
+    if (rfChannelIndex >= RF_CHANNEL_COUNT) {
+        rfChannelIndex = 0;
+    }
+    NRF24L01_SetChannel(rfChannels[rfChannelIndex]);
+}
+
+// The hopping channels are determined by the txId
+STATIC_UNIT_TESTED void cx10SetHoppingChannels(const uint8_t* txId)
+{
+    rfChannelIndex = 0;
+    rfChannels[0] = 0x03 + (txId[0] & 0x0F);
+    rfChannels[1] = 0x16 + (txId[0] >> 4);
+    rfChannels[2] = 0x2D + (txId[1] & 0x0F);
+    rfChannels[3] = 0x40 + (txId[1] >> 4);
+}
+
+/*
+ * This is called periodically by the scheduler.
+ * Returns NRF24_RECEIVED_DATA if a data packet was received.
+ */
+nrf24_received_t cx10DataReceived(uint8_t *payload)
+{
+    nrf24_received_t ret = NRF24_RECEIVED_NONE;
+    switch (protocolState) {
+    case STATE_BIND:
+        if (NRF24L01_ReadPayloadIfAvailable(payload, payloadSize + 2)) {
+            XN297_UnscramblePayload(payload, payloadSize + 2);
+            ret = NRF24_RECEIVED_BIND;
+            const bool bindPacket = checkBindPacket(payload);
+            if (bindPacket) {
+                // set the hopping channels as determined by the txId received in the bind packet
+                cx10SetHoppingChannels(txId);
+                ret = NRF24_RECEIVED_BIND;
+                protocolState = STATE_ACK;
+                ackCount = 0;
+            }
+        }
+        break;
+    case STATE_ACK:
+        // transmit an ACK packet
```

Comment:
There is no explicit delay between sending acks .. is sheduler frequency low enough?